stock_routing_operation/models/stock_move.py

```diff
@@ -0,0 +1,319 @@
+# Copyright 2019 Camptocamp (https://www.camptocamp.com)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+from itertools import chain
+
+from odoo import models
+
+
+class StockMove(models.Model):
+    _inherit = "stock.move"
+
+    def _action_assign(self):
+        # TODO use savepoint on the original assign instead of using
+        # unreserve
+        super()._action_assign()
+        if not self.env.context.get("exclude_apply_routing_operation"):
+            self._apply_src_move_routing_operation()
+            self._apply_dest_move_routing_operation()
+
+    def _apply_src_move_routing_operation(self):
+        src_moves = self._split_per_src_routing_operation()
+        src_moves._apply_move_location_src_routing_operation()
+
+    def _apply_dest_move_routing_operation(self):
+        dest_moves = self._split_per_dest_routing_operation()
+        dest_moves._apply_move_location_dest_routing_operation()
+
+    def _bypass_routing_operation_application(self, routing_type):
```

Comment:
@florian-dacosta FYI, I'll remove this in the next evolutions of the module: it's possible to filter the moves by inheriting `_apply_dest_move_routing_operation` and `_apply_src_move_routing_operation` and calling super only for the records you want, and it's more optimized because you can filter all the moves at once (`_bypass_routing_operation_application` is called record by record).

Also, I'll add an optional domain on the routing operation itself: if the domain returns no result for a move, the routing is not applied.

With these 2 possibilities, I'm sure you can handle your use case.

Reply:
@guewen 
Ok sure, I'll be able to manage it this way for sure, thanks.

I am actually working on it once more (still v8 thought !) and I have to manage another case...(a lot of exception you will tell me), but it still for the same problematic that, I want to apply the routing operation depending on the carrier (if any), that is my one big requirement that really make me differ on the cases handled by the module.

And so, I have a problem on this part : https://github.com/guewen/stock-logistics-warehouse/blob/13.0-add-stock_routing_operation/stock_routing_operation/models/stock_location.py#L90
You stop the loop on the first picking type found returning it.
In my case, the first one found may not be relevant, depending on my carrier. Then, I need to find the next one...etc...
There is not way for me to do it with the present module.
I see kind of 3 solutions
1) passing the move to the `_find_picking_type_for_routing` on location and add a hook in the loop to check if the found picking type is ok
2) just add a hook in the loop like `if match._is_allowed():` and manage my case with context and overriding `_is_allowed`
3) Add more logic in `_find_picking_type_for_routing` on `stock.move` and less on location side, but I don't think you will like that.

But anyway, I don't think it is worth proposing something about it now, because you say that the routing operation finding will change drastically in the neart future.
Beside, since I will definitly need the module for version 14 in the relatively near future, and I guess (hope) it won't be very different from v13, I can help with this refactore in v13 if needed.

Reply:
> @florian-dacosta FYI, I'll remove this in the next evolutions of the module: it's possible to filter the moves by inheriting _apply_dest_move_routing_operation and _apply_src_move_routing_operation and calling super only for the records you want, and it's more optimized because you can filter all the moves at once (_bypass_routing_operation_application is called record by record).

Following this logic, the check on move state (https://github.com/guewen/stock-logistics-warehouse/blob/13.0-add-stock_routing_operation/stock_routing_operation/models/stock_move.py#L49) should also be done in `_apply_src_move_routing_operation` right?

Reply:
> [](https://github.com/guewen/stock-logistics-warehouse/blob/13.0-add-stock_routing_operation/stock_routing_operation/models/stock_location.py#L90)
> You stop the loop on the first picking type found returning it.
> In my case, the first one found may not be relevant, depending on my carrier. Then, I need to find the next one...etc...

Should be supported with the change of data model we plan:

We'll have a model for routing with a location (M2o) field and routing rules, each rule having:

* sequence
* applied on source or destination
* picking type to apply
* domain

The first routing rule to match being applied.

Reply:
Well, that's nice, I look forward to see these  changes !
Once again, if I can be of any help, let me know. Else, I'll test and review when it is done